processor/src/memory/mod.rs

```diff
@@ -224,12 +224,34 @@ impl Memory {
         self.get_values(RangeInclusive::new(0, u64::MAX))
     }
 
-    /// Returns values within a range of addresses.
+    /// Returns values within a range of addresses at the last clock cycle.
     pub fn get_values(&self, range: RangeInclusive<u64>) -> Vec<(u64, Word)> {
+        self.get_values_at(range, None)
+    }
+
+    /// Returns values within a range of addresses, or optionally all values at a partciular step.
+    /// If no step is specified, then values at the last step are returned.
+    pub fn get_values_at(
+        &self,
+        range: RangeInclusive<u64>,
+        step: Option<usize>,
+    ) -> Vec<(u64, Word)> {
         let mut data: Vec<(u64, Word)> = Vec::new();
-        for (&addr, addr_trace) in self.trace.range(range) {
-            let value = addr_trace.last().expect("empty address trace").1;
-            data.push((addr, value));
+        if let Some(s) = step {
+            for (&addr, addr_trace) in self.trace.range(range) {
+                if let Some(t) = addr_trace
+                    .iter()
+                    .filter(|&x| x.0.as_int() <= s as u64)
+                    .max_by_key(|&x| x.0.as_int())
+                {
+                    data.push((addr, t.1));
+                }
```

Comment:
This search will probably result in linear complexity. But we know that addr_trace is sorted - so, we should be able to use [binary search](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.binary_search_by) here.

Reply:
fixed